### ISSUE TITLE
Design proposal: reliable delivery

### ISSUE THREAD

**Issue by hseeberger** [ockam-network/proposals#89]
*Status: draft, up for discussion*

Messages delivery across the network is not reliable, i.e. messages can be lost. For some use cases this cannot be accepted and at-least-once or effectively-once delivery semantics are needed. When talking about these, it is important to distinguish between the mere delivery and the complete processing of delivered messages: after a message has been successfully delivered across the network to some node the recipient worker might still crash before or while handling it.

Therefore we need to distinguish between two different possible flavors:
- In non-intrusive reliable delivery the recipient is not aware of the mechanisms in place and need not – even cannot – take special actions; this only allows for reliable delivery until the recipient node
- In confirmed reliable delivery the recipient explicitly needs to confirm the successful handling of each delivered message

## Common design approach

To keep things simple, the following will only look at pipes, i.e. reliable delivery from a sender to a receiver. The design can equally be applied to the reverse direction.

```
                   ┌───────────────┐  (msg, seq_no)  ┌───────────────┐
┌──────────┐  msg  │               │────────────────▶│               │  msg  ┌──────────┐
│  Sender  │──────▶│  SenderCntrl  │                 │ ReceiverCntrl │──────▶│ Receiver │
└──────────┘       │               │◀────────────────│               │       └──────────┘
                   └───────────────┘   ack(seq_no)   └───────────────┘
```

Both above flavors introduce two intermediaries between the sender and receiver. When the SenderController receives a message from the Sender, it first stores it locally, then forwards it alongside with an increasing sequence number to the ReceiverController; if it receives an acknowledgement from the ReceiverController it removes the stored message, if not it resends the message after some timeout.

## Flow control

In its simplest form only one message would ever be in flight between SenderCntrl and ReceiverCntrl. This could be increased, if the ReceiverCntrl also stored messages in order to be able to forward these to the Receiver in order. For both cases the local storage should be bounded, potentially resulting in dropping messages sent from the Sender to the SenderCntrl. This obviously violates the purpose of reliable delivery and hence should only be considered for low frequence senders.

```
┌──────────┐request┌───────────────┐
│          │◀──────│               │
│  Sender  │       │  SenderCntrl  │
│          │──────▶│               │
└──────────┘  msg  └───────────────┘
```

In order to avoid potential message dropping, instead of the Sender pushing messages to the SenderCntrl, the SenderCntrl has to pull from the Sender. This means introducing a protocol between the two which no longer is non-intrusive.

## Receiver confirmations

It might be necessary to retry delivery until the receiver has successfully handled a message.

```
┌──────────┐request┌───────────────┐  (msg, seq_no)  ┌───────────────┐  msg  ┌──────────┐
│          │◀──────│               │────────────────▶│               │──────▶│          │
│  Sender  │       │  SenderCntrl  │                 │ ReceiverCntrl │       │ Receiver │
│          │──────▶│               │◀────────────────│               │◀──────│          │
└──────────┘  msg  └───────────────┘   ack(seq_no)   └───────────────┘  ack  └──────────┘
```

This cannot happen non-intrusively, because it is up to the domain logic to confirm when exactly a message has been handled. Therefore we need to introduce another protocol between ReceiverCntrl and Receiver where the Receiver explicitly confirms back to the ReceiverCntrl. This also allows for end-to-end flow control driven by the Receiver.

**Comment by hairyhum** [comment]
@hseeberger 
In the flow control section, does that mean that the sender now has to store messages in a buffer and only send them to SenderCntrl when requested? Where would the application code live in this case and how does it interact with the buffer?

What's also interesting is how would SenderCntrl and ReceiverCntrl keep the connection between them alive (and keep each other accessible). Do you have any thoughts on how to implement that?

Also would it work if sender and receiver nodes are not connected directly but via cloud node like in [the forwarding example](https://github.com/ockam-network/ockam/blob/develop/examples/rust/get_started/examples/old/11-forwarding-via-a-cloud-node-responder.rs)?

**Comment by hseeberger** [comment]
> In the flow control section, does that mean that the sender now has to store messages in a buffer ... 

I think there are several options. One – maybe not suitable for Ockam channels – is to expose the pull approach throughout the sender. Another one could be to make sending an async operation which completes immediately once the message was stored in the buffer of the SenderCntrl, which means that this "blocks" if the buffer is full. In this approach it would not be possible to thread through the successfull delivery to the sender. That would only work if sending were an async operation which only completes once the message has been delivered (and acknowledged) by the receiver and this ACK has made its way back.

> What's also interesting is how would SenderCntrl and ReceiverCntrl keep the connection between them alive (and keep each other accessible). Do you have any thoughts on how to implement that?

I'd rather push that to the transport layer. One possibility that comes to my mind is heartbeats, failure detection and reconnects at the transport layer. But to my knowledge this is a serious challenge, in particular when dealing with not-super-short network partitions and the resulting task to decide if the other end has finally crashed (once it is concidered dead, it must not become a zombie again).

> Also would it work if sender and receiver nodes are not connected directly but via cloud node

Yes, I hope so, because I explicitly moved the logic for the reliable delivery to the application layer, so it should work end-to-end across any transports.

**Comment by hairyhum** [comment]
> One – maybe not suitable for Ockam channels – is to expose the pull approach throughout the sender
I'm not sure I quite understand what that means.
> That would only work if sending were an async operation which only completes once the message has been delivered (and acknowledged) by the receiver and this ACK has made its way back.
Is that going to utilize the Rust async-await mechanism?  Or by async you mean to have an ack handler in the sender?
> I'd rather push that to the transport layer. One possibility that comes to my mind is heartbeats, failure detection and reconnects at the transport layer. But to my knowledge this is a serious challenge
Yes. That's why dealing with it on the transport layer is challenging and might not work. Of course depends on what you mean by "transport layer". 
The [ockam routing protocol](https://github.com/ockam-network/proposals/tree/main/design/00070-routing-protocol) describes communication between endpoints over multiple transport connections. These connections may be unreliable and dynamically created.
In case when devices are connected via a cloud node or more complex chain of transports, how can we ensure that reliable connection (chain of transport connections) between those devices and SenderCnrtl<->ReceiverCnrtl workers is maintained?

**Comment by hseeberger** [comment]
> how can we ensure that reliable connection (chain of transport connections) between those devices and SenderCnrtl<->ReceiverCnrtl workers is maintained

That generally cannot be ensured. In any distributed sytem there is no way for one party to decide whether another party is "just" temporarily unavailable, e.g. because of network issues, or has permanently crashed. Therefore the only – to my knowledge – way to deal with failure of permantent connections is to use a failure detector (e.g. phi accrual detection) and "declare" the remote party as "dead" after some failure threashold. I think this shoud not be part of the application layer whereas reliable delivery (including acknowledgements) should be.

**Comment by hairyhum** [comment]
> I think this shoud not be part of the application layer whereas reliable delivery (including acknowledgements) should be.

I agree that for reliability reasons floating acknowledgements to the application layer is a good idea because it exposes the messaging limitations.

Although I think there is some misunderstanding regarding layers.

Messaging over Ockam Routing and Secure Channels are not on the Application Layer per se, as it's not always exposed to the user application and just carries user application data between endpoints. Neither it's the Transport Layer, because Ockam Routing is supposed to work over multiple transport hops.

In the OSI model it would be closer to what's called the Session Layer, making sure endpoints can discover and exchange messages with each other.  Somewhat similar to the role that other messaging systems (e.g. Kafka or AWS SNS or RabbitMQ) play in other service architectures.

Ockam is trying to build the Session Layer so the users can abstract away from some transport layer concerns like security or discoverability or reliability (although some aspects of those might float up into the Application Layer).

So I'm trying to raise the question on the Session Layer on how can we maintain the session which allows messaging, because the transport layer does not give us the session over multiple transport hops.
